modules/SceGxm: Improve sync object handling and simplify display queue logic

saturnsky opened the pull request:
This PR is intended to address issue #3403.

The logic of `sceGxmDisplayQueueAddEntry` has been modified according to the implementation of the PS Vita, so that it now waits for both `oldBufferSync` and `newBufferSync`. After this change, workaround codes that seemed unnecessary have also been removed. Additionally, the `handle_wait_sync_object` command has been removed, as it no longer appears logically necessary after these changes. Now, blocks related to `SyncObject` exist only within the logic of `sceGxmDisplayQueueAddEntry` (specifically in `display_entry_thread`).

This is a very aggressive modification, so there is a possibility of errors. However, based on my understanding, this change does not seem to introduce any logical issues.

The part related to frame prediction has been left unchanged as I couldn't fully assess the complexity. However, it may also be necessary to modify the frame prediction to target both the old and new buffers. This aspect is outside the scope of this PR.

Although I did not observe any regressions in my game library, I cannot be entirely certain that there are no issues due to the large scale of the changes. Additionally, there is a possibility that I misunderstood the existing lock logic and proceeded with development based on incorrect assumptions. Therefore, I am approaching this PR with caution.

Macdu commented:
You are removing the sync object waits on scenes and transfer gxm functions. This definitely loosk wrong: a scene/copy after a displayQueueAddEntry must wait for the sync object before starting .One of the goal of the sync object is not to draw to the surface being rendered (although this wouldn't happen anyway on Vita3K), also it is possible this can introduce race conditions for the rendering buffers.

saturnsky commented:
I understood it. I had removed the WaitSyncObject because I thought the handling would be explicitly done in sceGxmDisplayQueueAddEntry on the PS Vita, but it seems that change won't be beneficial. I reverted the changes related to WaitSyncObject and kept only the modifications in other parts.
I tested it within the same scope as when I previously submitted the PR, and within the range I tested, the new PR performed equivalently to the previous one. Additionally, the possibility of the side effects you were concerned about should also be reduced.

saturnsky commented:
My previous PR did not ensure that `sceGxmDisplayQueueAddEntry` would return only after the sync through `SceGxmSyncObject` had been completed. The current code more accurately emulates the actual behavior of `sceGxmDisplayQueueAddEntry` compared to before.

In fact, this revised code addresses a regression in [RecoLove: Blue Ocean](https://github.com/Vita3K/compatibility/issues/820).

However, it appears that the original Vita3K code aimed to minimize blocking on the main thread, while the revised PR introduces blocking on the main thread. Although I believe this PR is more accurate, I'm uncertain if it’s acceptable, as I lack sufficient historical information to make that determination.

Could you provide your thoughts on this aspect?

saturnsky commented:
Sorry for the confusion. I made a poor decision while developing in a fatigued state. It’s natural that making the logic synchronous would fix rendering issues in some games. 

However, it's correct for `sceGxmDisplayQueueAddEntry` not to block, and blocking should be handled within `display_entry_thread`. I should have remembered that just because more games function properly doesn't mean the issue was correctly identified. The previous PR was correct, and the cause of that game's malfunction likely lies elsewhere.

The PR has been reverted to its original state.

Macdu commented:
Indeed, the only time sceGxmDisplayQueueAddEntry should block is when the display queue is full. With the PR in its current state, it can still lead to race conditions (scenes began after sceGxmDisplayQueueAddEntry with the matching sync object should wait for the entry flip to be done).
What needs to be done for both the old and new sync objects is to:
- Wishlist at the timestamp when the entry was added
- Display the entry
- Call subject_done with the next timestamp

Also if the old and new sync objects are the same (this happens on some games), this should only be done once.

saturnsky commented:
Thank you for your continued advice.
I modified it to increment timestamp_ahead for both SyncObjects in sceGxmDisplayQueueAddEntry and to perform subject_done when display_thread->run_guest_function is called.
Come to think of it, with these changes, WaitSyncObject now has a meaningful purpose.

Macdu commented on vita3k/modules/SceGxm/SceGxm.cpp:
```diff
@@ -911,6 +910,8 @@ static void display_entry_thread(EmuEnvState &emuenv) {
 
+        // sceGxmDisplayQueueAddEntry waits for both buffers to complete
+        renderer::wishlist(old_sync, display_callback->old_sync_timestamp);
         renderer::wishlist(new_sync, display_callback->new_sync_timestamp);
```
You can add a check to perform the second wishlist only if old_sync != new_sync

saturnsky replied:
Okay. I didn’t add a comparative statement since I believed there was no functional difference between the two, however, considering it involves access to a lock, there may be a slight performance difference.

Macdu reviewed (approved):
lgtm, If possible, some testing would be nice to make sure this does not cause any regression.

nishinji commented:
No problems with games over 30.
